ansible-galaxy - only install/download collections which support ansible-core by default

s-hertel opened the pull request:
##### SUMMARY

Fixes #78539

* Make ansible-galaxy consider existing `COLLECTIONS_ON_ANSIBLE_VERSION_MISMATCH`  configuration. Now collections with mismatched requirements are only installed/downloaded if it's set to "ignore".

##### ISSUE TYPE

- Bugfix Pull Request

s-hertel committed:
ansible-galaxy - only install/download collections with compatible requires_ansible metadata

inject requires_ansible as a dependency of the collection

add a hack to preserve the collection origin in the error message in get_dependencies and find_matches

webknjaz commented:
#86187 will fix RHEL 10 + there's AZP API problems making now, making VM provisioning fail.

webknjaz commented:
/azp run

s-hertel committed:
yield requires_ansible requirement before dependencies

s-hertel committed:
Fix error handling and broken test

Fix getting requires_ansible dependency after the metadata is available

Allow 3rd party servers to provide this metadata lazily for individual versions

Rename the candidate attr name for the error handling hack

s-hertel committed:
fix diff

s-hertel committed:
clarify code comments

refactor a bit

s-hertel committed:
Don't duplicate error message. Before:

[ERROR]: Failed to resolve the requested dependencies map. Could not satisfy the following requirements:
* ns.col2:1.0.0 (dependency of ns.col3:1.0.0) requires ansible-core <2.19.1
Hint: To disregard whether the collection supports the current version of ansible-core, configure COLLECTIONS_ON_ANSIBLE_VERSION_MISMATCH as "ignore".
Hint: Pre-releases hosted on Galaxy or Automation Hub are not installed by default unless a specific version is requested. To enable pre-releases globally, use --pre: [RequirementInformation(requirement=<ansible-core:<2.19.1 of type 'requires_ansible' from Galaxy>, parent=<ns.col2:1.0.0 of type 'galaxy' from cmd_arg>)]

Failed to resolve the requested dependencies map. Could not satisfy the following requirements:
* ns.col2:1.0.0 (dependency of ns.col3:1.0.0) requires ansible-core <2.19.1
Hint: To disregard whether the collection supports the current version of ansible-core, configure COLLECTIONS_ON_ANSIBLE_VERSION_MISMATCH as "ignore".
Hint: Pre-releases hosted on Galaxy or Automation Hub are not installed by default unless a specific version is requested. To enable pre-releases globally, use --pre.

<<< caused by >>>

[RequirementInformation(requirement=<ansible-core:<2.19.1 of type 'requires_ansible' from Galaxy>, parent=<ns.col2:1.0.0 of type 'galaxy' from cmd_arg>)]

After:

[ERROR]: Failed to resolve the requested dependencies map. Could not satisfy the following requirements:
* ns.col2:1.0.0 (dependency of ns.col3:1.0.0) requires ansible-core <2.19.1
Hint: To disregard whether the collection supports the current version of ansible-core, configure COLLECTIONS_ON_ANSIBLE_VERSION_MISMATCH as "ignore".
Hint: Pre-releases hosted on Galaxy or Automation Hub are not installed by default unless a specific version is requested. To enable pre-releases globally, use --pre.

s-hertel committed:
appease CI

fix type hints

Shrews commented on changelogs/fragments/install-ansible-core-compatible-collections.yml:
```diff
@@ -0,0 +1,6 @@
+bugfixes:
+- >-
+  ``ansible-galaxy install`` and ``ansible-galaxy collection install|download` - now consider
```
```suggestion
  ``ansible-galaxy install`` and ``ansible-galaxy collection install|download`` - now consider
```

sivel replied:
```suggestion
  ``ansible-galaxy install`` and ``ansible-galaxy collection install|download`` - now consider
```

missing backtick

s-hertel committed:
Update changelogs/fragments/install-ansible-core-compatible-collections.yml

Co-authored-by: sivel / Matt Martz <matt@sivel.net>

sivel commented on test/integration/targets/ansible-galaxy-collection/vars/main.yml:
```diff
@@ -201,3 +201,19 @@ collection_list:
+    name: name_requires_ansible
+    version: 1.0.0
+    runtime: |-
+      requires_ansible: "<{{ ansible_version.major }}.{{ ansible_version.minor }}"
```
```suggestion
      requires_ansible: "<1.1"
```

Can we make this static?  And picking something like `1.1` is meaningless enough but will also cause the tests to trip in the right way.

The work I am doing to split `setup_collections` out of the actual testing, means that the ansible version used to generate the collections in the container may get out of sync with devel.

s-hertel replied:
Good idea. I've fixed these.

sivel commented on test/integration/targets/ansible-galaxy-collection/tasks/check_requires_ansible.yml:
```diff
@@ -0,0 +1,126 @@
+    install_or_download: "ansible-galaxy collection {{ subcommand }}{{ (subcommand == 'install') | ternary(' --force-with-deps', '')}}"
+    failed_error: |
+        \[ERROR\]: Failed to resolve the requested dependencies map\. Could not satisfy the following requirements:
+        \* ns_requires_ansible\.name_requires_ansible:1\.0\.0 \(direct request\) requires ansible-core <{{ ansible_version.major }}\.{{ ansible_version.minor }}
```
```suggestion
        \* ns_requires_ansible\.name_requires_ansible:1\.0\.0 \(direct request\) requires ansible-core <1\.1
```

sivel commented on test/integration/targets/ansible-galaxy-collection/tasks/check_requires_ansible.yml:
```diff
@@ -0,0 +1,126 @@
+        shell: >-
+          echo 'requires_ansible: "{{ incompatible_version }}"' > {{ temp2 }}/ansible_collections/ns_requires_ansible/name_requires_ansible/meta/runtime.yml
+        vars:
+          incompatible_version: "<{{ ansible_version.major }}.{{ ansible_version.minor }}"
```
```suggestion
          incompatible_version: "<1.1"
```

sivel commented on test/integration/targets/ansible-galaxy-collection/tasks/check_requires_ansible.yml:
```diff
@@ -0,0 +1,126 @@
+    vars:
+      expected_message: |
+        \[ERROR\]: Failed to resolve the requested dependencies map\. Could not satisfy the following requirements:
+        \* ns_requires_ansible\.name_requires_ansible:1\.0\.0 \(dependency of ns_requires_ansible\.dependency_requires_ansible:2\.0\.0\) requires ansible-core <{{ ansible_version.major }}\.{{ ansible_version.minor }}
```
```suggestion
        \* ns_requires_ansible\.name_requires_ansible:1\.0\.0 \(dependency of ns_requires_ansible\.dependency_requires_ansible:2\.0\.0\) requires ansible-core <1\.1
```

sivel reviewed (changes requested):
A recommendation for future maintainability of the tests

s-hertel committed:
Hardcode unsupported requires_ansible version to make maintaining the tests easier

Add a test for failed backtracking

Co-authored-by: sivel / Matt Martz <matt@sivel.net>

sivel commented:
> @sivel I tried to move the `get_dependencies` tracking into `CollectionDependencyReporter.adding_requirement`, but abandoned it because it can cause verbose, confusing error messages when backtracking. If there's a way to prune the requirements originating from potential candidates that were discarded, it could work... Stashing the data on Requirement/Candidate objects avoids conflating the current dependency chain with backtracked dependency chains containing the same FQCN. I added test coverage for requires_ansible failures after backtracking.

Understood. Thank you for looking into this in more depth. I don't have a great solution in mind immediately to resolve this, and I think we can leave it for a later change if we can ever get around to a different implementation. But I don't think we should continue blocking the PR for that.

felixfontein commented on changelogs/fragments/install-ansible-core-compatible-collections.yml:
```diff
@@ -0,0 +1,6 @@
+bugfixes:
```
Shouldn't there also be a `minor_changes` or even `major_changes` entry in the changelog for this? After all, this is quite a major new feature, and not just a bugfix.

s-hertel replied:
It seems like a bugfix to me. ansible-galaxy shouldn't install collections that would bomb out or give warnings at runtime by default. What kind of issues do you see with backporting this? I can't imagine why anyone would want the previous behavior by default.

s-hertel replied:
Opened a backport https://github.com/ansible/ansible/pull/86624, though I don't know if the release managers will accept it. Would love to hear your opinion too.

felixfontein replied:
To me this is not a bugfix since it's a breaking change in behavior. Just because collections require a certain ansible-core version doesn't mean they (or parts of them) don't work with older ansible-core versions. So backporting this might break existing uses of ansible-core 2.20 that are working. (Obviously this currently causes warnings, but not everyone minds warnings, some people just ignore them as long as it works ;-) )

I totally agree that this should have been the default behavior for a longer time, I just don't think it's a good idea to retroactive introduce it into older versions since the current behavior has been the default for a long time - basically since `requires_ansible` was introduced in ansible-base 2.10.

The changelog entry should be more clear about what the result of this change is. For most users, from the text above it's unclear that this no longer installs collections that don't support the current ansible-core version, and they would expect this to be announced as at least a new feature, or even as a breaking change.

s-hertel replied:
I think that is complete a misuse of the metadata. If a collection documents that it doesn't support the current version of ansible-core but actually does, runtime has no way to determine that. The previous behavior of `ansible-galaxy` would blindly install that collection, but subsequently listing or resinstalling would give a warning or error. If https://docs.ansible.com/projects/ansible/latest/reference_appendices/config.html#collections-on-ansible-version-mismatch is set to `error`, there would be no way for the user to fix that via ansible-galaxy - they'd have to manually delete that collection so it could be reinstalled, or remove the configuration.

s-hertel replied:
I've opened a PR to try to rephrase the changelog https://github.com/ansible/ansible/pull/86626. I haven't changed the section (yet, anyway).
